File: lib/go/adapters/interceptor/from_yaml.go
package radixipgrpc

// from_yaml.go — hot-reload integration for gRPC via fsnotify.
//
// Usage:
//
//	interceptor, stop, err := radixipgrpc.NewFromYAML("radixip.yaml", engineAdapter)
//	if err != nil { log.Fatal(err) }
//	defer stop()
//	s := grpc.NewServer(
//	    grpc.UnaryInterceptor(interceptor),
//	    grpc.StreamInterceptor(interceptor),
//	)

import (
	"context"
	"log"
	"sync/atomic"

	"github.com/Mwangi-Derrick/radixip/lib/go/config"
	"github.com/Mwangi-Derrick/radixip/lib/go/policy"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
)

// yamlState bundles a config snapshot with the limiter derived from it.
type yamlState struct {
	cfg     *config.RadixIpConfig
	limiter *policy.TokenBucketLimiter
}

func newYAMLState(cfg *config.RadixIpConfig) *yamlState {
	rl := cfg.RadixIP.RateLimit
	lim := policy.NewTokenBucketLimiter(
		rl.Capacity,
		rl.RefillRate,
		rl.TTLSeconds,
		rl.MaxBuckets,
	)
	log.Printf("radixipgrpc: (re)built limiter capacity=%d refill=%d/s", rl.Capacity, rl.RefillRate)
	return &yamlState{cfg: cfg, limiter: lim}
}

// grpcWatcher holds the atomic hot-swap state.
type grpcWatcher struct {
	watcher *config.Watcher
	state   atomic.Pointer[yamlState]
	engine  Engine
}

// UnaryInterceptor implements the gRPC unary interceptor for the watcher.
func (g *grpcWatcher) UnaryInterceptor() grpc.UnaryServerInterceptor {
	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
		latest := g.watcher.Current()
		s := g.state.Load()

		// Config pointer changed → rebuild limiter.
		if s.cfg != latest {
			next := newYAMLState(latest)
			g.state.Store(next)
			s = next
		}

		mwCfg := latest.RadixIP.Middleware
		trusted := parseCIDRs(mwCfg.TrustedProxies)

		ip, err := extractIPFromContext(ctx, trusted)
		if err != nil {
			return nil, status.Errorf(codes.InvalidArgument, "failed to extract IP: %v", err)
		}
		ipStr := ip.String()

		// Blocklist check.
		if latest.RadixIP.Blocklist.Enabled && g.engine != nil {
			if g.engine.Lookup(ipStr) {
				return nil, status.Errorf(
					statusCodeToGRPC(mwCfg.Responses.Blocked),
					"blocked: IP %s is in blocklist",
					ipStr,
				)
			}
		}

		// Rate limit check.
		if latest.RadixIP.RateLimit.Enabled && s.limiter != nil {
			key := bucketKey(ip, latest.RadixIP.RateLimit.BucketMode.Mode)
			if !s.limiter.Allow(key) {
				// Add retry-after to response metadata
				if err := grpc.SetHeader(ctx, metadata.Pairs("retry-after", "1")); err != nil {
					// Log but continue
				}
				return nil, status.Errorf(
					statusCodeToGRPC(mwCfg.Responses.RateLimited),
					"rate limited: IP %s exceeded rate limit",
					ipStr,
				)
			}
		}

		return handler(ctx, req)
	}
}

// StreamInterceptor implements the gRPC stream interceptor for the watcher.
func (g *grpcWatcher) StreamInterceptor() grpc.StreamServerInterceptor {
	return func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
		ctx := ss.Context()
		latest := g.watcher.Current()
		s := g.state.Load()

		// Config pointer changed → rebuild limiter.
		if s.cfg != latest {
			next := newYAMLState(latest)
			g.state.Store(next)
			s = next
		}

		mwCfg := latest.RadixIP.Middleware
		trusted := parseCIDRs(mwCfg.TrustedProxies)

		ip, err := extractIPFromContext(ctx, trusted)
		if err != nil {
			return status.Errorf(codes.InvalidArgument, "failed to extract IP: %v", err)
		}
		ipStr := ip.String()

		// Blocklist check.
		if latest.RadixIP.Blocklist.Enabled && g.engine != nil {
			if g.engine.Lookup(ipStr) {
				return status.Errorf(
					statusCodeToGRPC(mwCfg.Responses.Blocked),
					"blocked: IP %s is in blocklist",
					ipStr,
				)
			}
		}

		// Rate limit check.
		if latest.RadixIP.RateLimit.Enabled && s.limiter != nil {
			key := bucketKey(ip, latest.RadixIP.RateLimit.BucketMode.Mode)
			if !s.limiter.Allow(key) {
				// Add retry-after to response metadata
				if err := grpc.SetHeader(ctx, metadata.Pairs("retry-after", "1")); err != nil {
					// Log but continue
				}
				return status.Errorf(
					statusCodeToGRPC(mwCfg.Responses.RateLimited),
					"rate limited: IP %s exceeded rate limit",
					ipStr,
				)
			}
		}

		return handler(srv, ss)
	}
}

// NewFromYAML creates a hot-reloading gRPC interceptor from a YAML config file.
//
// The engine adapter is passed separately because the ART tree state (blocklist
// data) is managed outside the config lifecycle. Only rate-limit parameters and
// interceptor options are hot-swapped when the file changes.
//
// stop() must be called on server shutdown to release fsnotify resources.
//
// Returns a unary interceptor, stream interceptor, and a stop function.
func NewFromYAML(path string, engine Engine) (grpc.UnaryServerInterceptor, grpc.StreamServerInterceptor, func(), error) {
	w, err := config.NewWatcher(path)
	if err != nil {
		return nil, nil, nil, err
	}

	gw := &grpcWatcher{watcher: w, engine: engine}
	gw.state.Store(newYAMLState(w.Current()))

	return gw.UnaryInterceptor(), gw.StreamInterceptor(), w.Stop, nil
}

// Alternative: Combined interceptor that returns a single interceptor that works
// for both unary and stream methods. This is useful if you want to use the same
// interceptor for both types.
type CombinedInterceptor interface {
	UnaryInterceptor() grpc.UnaryServerInterceptor
	StreamInterceptor() grpc.StreamServerInterceptor
}

// NewCombinedFromYAML returns a CombinedInterceptor interface that provides
// both unary and stream interceptors from the same YAML config.
func NewCombinedFromYAML(path string, engine Engine) (CombinedInterceptor, func(), error) {
	w, err := config.NewWatcher(path)
	if err != nil {
		return nil, nil, err
	}

	gw := &grpcWatcher{watcher: w, engine: engine}
	gw.state.Store(newYAMLState(w.Current()))

	return gw, w.Stop, nil
}

// // Helper function to convert HTTP status to gRPC status code.
// func statusCodeToGRPC(httpStatus int) codes.Code {
// 	switch httpStatus {
// 	case 403:
// 		return codes.PermissionDenied
// 	case 429:
// 		return codes.ResourceExhausted
// 	case 400:
// 		return codes.InvalidArgument
// 	default:
// 		return codes.Unknown
// 	}
// }

File: lib/rust/adapters/interceptor/src/from_yaml.rs
//! Hot-reloading gRPC interceptor for RadixIP.
//!
//! Watches a YAML configuration file and atomically hot-swaps rate-limit
//! parameters and middleware options when the file changes (zero downtime).
//!
//! # Usage
//!
//! ```rust,ignore
//! use radixip_grpc_interceptor::from_yaml::{GrpcWatchedRadixIpInterceptor, GrpcWatchedRadixIpLayer};
//! use radixip_policy::ConfigWatcher;
//! use std::sync::Arc;
//!
//! let watcher = Arc::new(ConfigWatcher::new("radixip.yaml")?);
//! let engine = Arc::new(my_radix_engine); // Box<dyn RadixEngine>
//!
//! // Option A — tonic Interceptor trait (metadata-only, cheapest)
//! let interceptor = GrpcWatchedRadixIpInterceptor::new(watcher.clone(), engine.clone());
//! let svc = tonic::service::interceptor(my_service, interceptor);
//!
//! // Option B — Tower Layer (full HTTP/2 request)
//! let layer = GrpcWatchedRadixIpLayer::new(watcher, engine);
//! let server = tonic::transport::Server::builder()
//!     .layer(layer)
//!     .add_service(my_service)
//!     .serve(addr)
//!     .await?;
//! ```

use std::sync::Arc;
use std::task::{Context, Poll};

use futures_util::future::BoxFuture;
use http::{Request, Response};
use std::str::FromStr;
use tonic::{metadata::MetadataValue, service::Interceptor, Status};
use tower::{Layer, Service};

use radixip::RadixEngine;
use radixip_policy::{extract_ip, ConfigWatcher};

// Interceptor with Hot-Reload (tonic::service::Interceptor)

/// Tonic `Interceptor` that hot-reloads its config from a YAML watcher.
///
/// Every RPC reads the current config snapshot via an atomic pointer load
/// (wait-free). When the file changes, the watcher rebuilds the policy state
/// in the background and the next call picks it up automatically.
#[derive(Clone)]
pub struct GrpcWatchedRadixIpInterceptor {
    watcher: Arc<ConfigWatcher>,
    engine: Arc<Box<dyn RadixEngine>>,
}

impl GrpcWatchedRadixIpInterceptor {
    /// Create a new hot-reloading RadixIP interceptor.
    pub fn new(watcher: Arc<ConfigWatcher>, engine: Arc<Box<dyn RadixEngine>>) -> Self {
        Self { watcher, engine }
    }
}

impl Interceptor for GrpcWatchedRadixIpInterceptor {
    fn call(&mut self, req: tonic::Request<()>) -> Result<tonic::Request<()>, Status> {
        let state = self.watcher.state(); // Wait-free atomic pointer load
        let mw_cfg = &state.config.radixip.middleware;
        let rl_cfg = &state.config.radixip.rate_limit;
        let bl_cfg = &state.config.radixip.blocklist;

        let metadata = req.metadata();

        // 1. Extract client IP from metadata headers.
        let xff = metadata
            .get("x-forwarded-for")
            .and_then(|v| v.to_str().ok());
        let x_real_ip = metadata.get("x-real-ip").and_then(|v| v.to_str().ok());
        // tonic sets remote_addr via Request::remote_addr()
        let remote_addr = req.remote_addr();

        let ip = match extract_ip(
            xff,
            x_real_ip,
            remote_addr.as_ref(),
            &mw_cfg.trusted_proxies,
        ) {
            Ok(ip) => ip,
            Err(e) => {
                return Err(Status::invalid_argument(format!(
                    "failed to extract IP: {}",
                    e
                )));
            }
        };

        // 2. Blocklist check.
        if bl_cfg.enabled && self.engine.lookup(&ip).is_some() {
            return Err(Status::permission_denied("blocked: IP is in blocklist"));
        }

        // 3. Rate limit check.
        if rl_cfg.enabled && !state.limiter.allow(ip, Some(self.engine.as_ref().as_ref())) {
            let mut status = Status::resource_exhausted("rate limited: exceeded rate limit");
            let retry_after =
                MetadataValue::from_str("1").unwrap_or_else(|_| MetadataValue::from_static("1"));
            status.metadata_mut().insert("retry-after", retry_after);
            return Err(status);
        }

        Ok(req)
    }
}

// Tower Layer with Hot-Reload (http::Request<B> level)

/// Tower `Layer` that applies hot-reloading RadixIP checks to every HTTP/2 request.
///
/// Prefer this over `GrpcWatchedRadixIpInterceptor` when you need access to
/// request extensions (e.g., `SocketAddr` injected by `TcpIncoming`) or when
/// composing with other Tower middleware.
#[derive(Clone)]
pub struct GrpcWatchedRadixIpLayer {
    watcher: Arc<ConfigWatcher>,
    engine: Arc<Box<dyn RadixEngine>>,
}

impl GrpcWatchedRadixIpLayer {
    /// Create a new hot-reloading RadixIP layer.
    pub fn new(watcher: Arc<ConfigWatcher>, engine: Arc<Box<dyn RadixEngine>>) -> Self {
        Self { watcher, engine }
    }
}

impl<S> Layer<S> for GrpcWatchedRadixIpLayer {
    type Service = GrpcWatchedRadixIpService<S>;

    fn layer(&self, service: S) -> Self::Service {
        GrpcWatchedRadixIpService {
            inner: service,
            watcher: self.watcher.clone(),
            engine: self.engine.clone(),
        }
    }
}

/// Tower service produced by `GrpcWatchedRadixIpLayer`.
#[derive(Clone)]
pub struct GrpcWatchedRadixIpService<S> {
    inner: S,
    watcher: Arc<ConfigWatcher>,
    engine: Arc<Box<dyn RadixEngine>>,
}

impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for GrpcWatchedRadixIpService<S>
where
    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Clone + Send + 'static,
    S::Future: Send + 'static,
    S::Error: Send + 'static,
    ReqBody: Send + 'static,
    ResBody: From<String> + Send + 'static,
{
    type Response = S::Response;
    type Error = S::Error;
    type Future = BoxFuture<'static, Result<Self::Response, Self::Error>>;

    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
        self.inner.poll_ready(cx)
    }

    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
        let mut inner = self.inner.clone();
        let watcher = self.watcher.clone();
        let engine = self.engine.clone();

        // Extract headers before moving req into the async block.
        let xff = req
            .headers()
            .get("x-forwarded-for")
            .and_then(|v| v.to_str().ok())
            .map(|s| s.to_string());
        let x_real_ip = req
            .headers()
            .get("x-real-ip")
            .and_then(|v| v.to_str().ok())
            .map(|s| s.to_string());
        let remote_addr = req.extensions().get::<std::net::SocketAddr>().copied();

        Box::pin(async move {
            let state = watcher.state(); // Wait-free atomic pointer load
            let mw_cfg = &state.config.radixip.middleware;
            let rl_cfg = &state.config.radixip.rate_limit;
            let bl_cfg = &state.config.radixip.blocklist;
            let responses = &mw_cfg.responses;

            // 1. Extract IP.
            let ip = match extract_ip(
                xff.as_deref(),
                x_real_ip.as_deref(),
                remote_addr.as_ref(),
                &mw_cfg.trusted_proxies,
            ) {
                Ok(ip) => ip,
                Err(e) => {
                    let body = ResBody::from(format!(r#"{{"error":"bad request: {}"}}"#, e));
                    let mut res = Response::new(body);
                    *res.status_mut() = http::StatusCode::BAD_REQUEST;
                    res.headers_mut().insert(
                        http::header::CONTENT_TYPE,
                        http::header::HeaderValue::from_static("application/json"),
                    );
                    return Ok(res);
                }
            };

            // 2. Blocklist check.
            if bl_cfg.enabled && engine.lookup(&ip).is_some() {
                let status = http::StatusCode::from_u16(responses.blocked)
                    .unwrap_or(http::StatusCode::FORBIDDEN);
                let body = ResBody::from(r#"{"error":"blocked"}"#.to_string());
                let mut res = Response::new(body);
                *res.status_mut() = status;
                res.headers_mut().insert(
                    http::header::CONTENT_TYPE,
                    http::header::HeaderValue::from_static("application/json"),
                );
                return Ok(res);
            }

            // 3. Rate limit check.
            if rl_cfg.enabled && !state.limiter.allow(ip, Some(engine.as_ref().as_ref())) {
                let status = http::StatusCode::from_u16(responses.rate_limited)
                    .unwrap_or(http::StatusCode::TOO_MANY_REQUESTS);
                let body = ResBody::from(r#"{"error":"rate limited"}"#.to_string());
                let mut res = Response::new(body);
                *res.status_mut() = status;
                res.headers_mut().insert(
                    http::header::CONTENT_TYPE,
                    http::header::HeaderValue::from_static("application/json"),
                );
                res.headers_mut().insert(
                    http::header::RETRY_AFTER,
                    http::header::HeaderValue::from_static("1"),
                );
                return Ok(res);
            }

            inner.call(req).await
        })
    }
}
